sql: support to_json of JSON builtins

heyihong opened the pull request:
Hi @justinj , 

I am still working on some AsJSON functions and tests. I am not sure if I am on the right track. Could you take a skim on it? 

Also, when I tried to refer [Postgres json tests](https://github.com/postgres/postgres/blob/b574228715f0fd77ed1f4f084603cff9e757e992/src/test/regress/expected/json.out) to write some tests, I found that the parser could not handle some of them successfully, like select date 'Infinity',SELECT to_json('{{1,5},{99,100}}'::int[]) and so on. I am not sure if they need to be taken into account. 

Related to #20120.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20234)
<!-- Reviewable:end -->

justinj commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -2310,6 +2314,19 @@ var jsonTypeOfImpl = tree.Builtin{
+		}
+		return tree.NewDJSON(j), nil
+	},
+	Info: "Returns the value as json or jsonb.",
```
nit: I think we usually capitalize type names in `Info`, though I don't think we're super consistent about it.

justinj commented on pkg/sql/sem/tree/datum.go:
```diff
@@ -273,6 +274,11 @@ func (*DBool) ResolvedType() types.T {
 	return types.Bool
 }
 
+// AsJSON implements transformation to JSON
```
Two style nits: we usually put periods after comments, and also we usually write these comments in the form `AsJSON implements the json.JSONizable interface`.

In addition, the specificity here is nice, but in the interest of keeping the interface of the `json` package smaller where possible, maybe it would be easier to just use `json.MakeJSON` here and in other methods where possible? It will be a little slower because of the switch, I think, but that's something that can be changed if we run into problems with it later. You might be able to even just add switch cases for things like `time` and `timeofday`?

justinj commented on pkg/sql/sem/tree/datum.go:
```diff
@@ -901,6 +927,11 @@ func NewDCollatedString(
 
+// AsJSON implements transformation to JSON
+func (d *DCollatedString) AsJSON() (json.JSON, error) {
+	return json.MakeJSONFromString(""), nil
```
This needs to work, but I think you can just use the string value, collation just affects sorting within SQL so I don't think there's anything special that needs to happen here.

justinj commented on pkg/sql/sem/tree/datum.go:
```diff
@@ -985,6 +1016,11 @@ func (*DBytes) ResolvedType() types.T {
 }
 
+// AsJSON implements transformation to JSON
+func (d *DBytes) AsJSON() (json.JSON, error) {
```
This one will require some tests to ensure the escaping for various characters (like non-printing and valid/invalid unicode characters) matches Postgres' behaviour.

justinj commented on pkg/util/json/json.go:
```diff
@@ -453,68 +551,25 @@ func (j jsonObject) EncodeInvertedIndexKeys(b []byte) [][]byte {
 func MakeJSON(d interface{}) (JSON, error) {
```
Nice cleanup to this function!!

justinj commented on pkg/sql/sem/tree/datum.go:
```diff
@@ -2701,6 +2806,38 @@ func (t *DTable) ResolvedType() types.T {
 }
 
+// AsJSON implements transformation to JSON
+func (t *DTable) AsJSON() (json.JSON, error) {
```
This doesn't seem to be tested, and I can't seem to find the corresponding docs on how this works in PG? Can you point me to it (or show me an example?)

justinj commented on pkg/util/json/json.go:
```diff
@@ -442,6 +447,99 @@ func (j jsonObject) EncodeInvertedIndexKeys(b []byte) [][]byte {
+}
+
+// makeJSONFromArrayOfInterface returns a JSON value given a []interface{}.
+func makeJSONFromArrayOfInterface(v []interface{}) (JSON, error) {
```
Since this function is unexported I think we can get away with a slightly less verbose name, maybe just `fromArray`?

justinj commented on pkg/util/json/json.go:
```diff
@@ -442,6 +447,99 @@ func (j jsonObject) EncodeInvertedIndexKeys(b []byte) [][]byte {
+}
+
+// makeJSONFromArrayOfInterface returns a JSON value given a map[string]interface{}.
+func makeJSONFromMapOfInterface(v map[string]interface{}) (JSON, error) {
```
same here

justinj commented on pkg/sql/sem/tree/datum.go:
```diff
@@ -71,6 +71,7 @@ var (
 // Datum represents a SQL value.
 type Datum interface {
 	TypedExpr
+	json.JSONizable
```
Consider adding a comment above this explaining that it's needed for the `to_jsonb` function, which needs to be able to convert arbitrary datums into JSON.

justinj commented on pkg/util/json/json.go:
```diff
@@ -442,6 +447,99 @@ func (j jsonObject) EncodeInvertedIndexKeys(b []byte) [][]byte {
+}
+
+// MakeJSONFromArrayOfJSON returns a JSON value given a []JSON.
+func MakeJSONFromArrayOfJSON(v []JSON) JSON {
```
I think you might have mentioned this before but I didn't understand at the time - I think you can get away without making this function if you add a check at the end of `MakeJSON` that just returns `JSON` implementors unchanged (#20217 also adds this functionality).

justinj reviewed (commented):
Hi @heyihong! Sorry for the delay, as we discussed before it's been a long weekend here. This is a really nice PR! I haven't compared your output against Postgres so I'll have to do another pass later once you've finished to ensure all of our conversions match there, but I like the overall approach you've taken.

It seems a bit noisy that every datum in `datum.go` is now concerned with its `AsJSON` impl, those have generally been much more general. It might make sense to put all of those in a new file, though I'm not super wedded to that idea, since they *are* part of the required interface now.

The general approach looks pretty solid to me but since it's quite nontrivial I'm going to ask @jordanlewis to take a look as well.

knz reviewed (changes requested):
Hi @heyihong 
Thank you for tackling this!

In general like Justin I think that adding a new method to `Datum` is not the way to go here. There are two arguments:

1. over time we'd like to see fewer methods over `Datum`, not more
2. there are many instances of the new `AsJSON` method that return the same result (empty string), so for the sake of DRY I would recommend them to be shared.

This suggests a single function (in a new file!) with the following structure:

```go
func AsJSON(d Datum) (json.JSON, error) {
   switch t := d.(type) {
      case *DTuple, *Placeholder, *DUuid, ...: 
           return json.MakeJSONFromString(""), nil
      case *DBytes:
           return json.MakeJSONFromString(string(*t)), nil
       ...
   }
}
```

justinj commented:
@knz I see where you're coming from but this direction makes me somewhat nervous - it seems we're giving up compiler hand-holding when adding a new Datum for not a whole lot of benefit. Obviously in an ideal world our programming language could check that such a switch is exhaustive and I would agree that that is the right choice, but as it is I feel like giving up any guard rails for making sure such a function is complete a little worrying.

Although, one thing we can do to make that slightly less concerning is that we can make a test that uses `RandDatum` and calls its `AsJSON` and just ensures that it returns something at all.

knz commented:
> I see where you're coming from but this direction makes me somewhat nervous - it seems we're giving up compiler hand-holding when adding a new Datum for not a whole lot of benefit.

I think this line of reasoning has to go. If this "hand holding" is really a concern we can write a test which generates random values of each type and tries all the APIs on it.

heyihong commented:
@justinj , @knz Thanks for the reviews. I will move all AsJSON function to one function, though I think putting AsJSON in Datum interface could help us discover the error earlier when adding a new Datum, compared with random tests. Also, I should replace most of AsJSON functions that return an empty string to the actual implementation. I still need to take more time to have them complete and tested.

heyihong commented:
Could I get some hints for implementing json serialization of Duuid, DOid, DInterval and DByte? For DUuid and DOid, I am not sure how to test it. For DInterval and DByte, I felt that I may need to rewrite the ways of serialization to string.

---

Review status: 0 of 6 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[pkg/sql/sem/builtins/builtins.go, line 2327 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-Kzb292Sc_Bw_bUuZ4UT-r1-2327:-L-5HfCsN0d9x1F1q_dx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fb3b2fc9ab4ef868f74a0d0070cd8e5d5d3b247/pkg/sql/sem/builtins/builtins.go#L2327)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

nit: I think we usually capitalize type names in `Info`, though I don't think we're super consistent about it.
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/as_json.go, line 57 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-L-5KyhV1rnJ6J1uLeQ3:-L-5KyhV1rnJ6J1uLeQ4:b-gpct7d) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4aa2f81c45f95baf15fb784c3639b8cfc86fd80/pkg/sql/sem/tree/as_json.go#L57)):*
> ```Go
> 		return AsJSON(t.Wrapped)
> 	}
> 	return nil, nil
> ```

I should return an error at the end of the code. Could I return like this pgerror.NewErrorf(pgerror.CodeInternalError, "unexpected datum type %T", d.(type))?

---

*[pkg/sql/sem/tree/datum.go, line 74 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-Kzb292Sc_Bw_bUuZ4UU-r1-74:-L-5HmxFa8Kmkv8LhnFx:b-auzin5) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fb3b2fc9ab4ef868f74a0d0070cd8e5d5d3b247/pkg/sql/sem/tree/datum.go#L74)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Consider adding a comment above this explaining that it's needed for the `to_jsonb` function, which needs to be able to convert arbitrary datums into JSON.
</blockquote></details>

Changed it to a single AsJSON method.

---

*[pkg/sql/sem/tree/datum.go, line 277 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-Kzb292Sc_Bw_bUuZ4UU-r1-277:-L-5HytuVLxC9bHkI9ar:b5idrsd) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fb3b2fc9ab4ef868f74a0d0070cd8e5d5d3b247/pkg/sql/sem/tree/datum.go#L277)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Two style nits: we usually put periods after comments, and also we usually write these comments in the form `AsJSON implements the json.JSONizable interface`.

In addition, the specificity here is nice, but in the interest of keeping the interface of the `json` package smaller where possible, maybe it would be easier to just use `json.MakeJSON` here and in other methods where possible? It will be a little slower because of the switch, I think, but that's something that can be changed if we run into problems with it later. You might be able to even just add switch cases for things like `time` and `timeofday`?
</blockquote></details>

Adding more switch cases will be more convenient to use. But I am concerning that using MakeJSON skips type check which makes type errors a little difficult to discover after some changes and because there is no standard way to serialize time and timeofday, adding switch cases seems to give more restrictions about JSON serialization of time and timeofday.

---

*[pkg/sql/sem/tree/datum.go, line 932 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-Kzb292Sc_Bw_bUuZ4UU-r1-932:-L-5JMsqYbDfpTrUw9G_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fb3b2fc9ab4ef868f74a0d0070cd8e5d5d3b247/pkg/sql/sem/tree/datum.go#L932)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

This needs to work, but I think you can just use the string value, collation just affects sorting within SQL so I don't think there's anything special that needs to happen here.
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/datum.go, line 1020 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-Kzb292Sc_Bw_bUuZ4UU-r1-1020:-L-5JQunFlgG0P5PBRTh:bgxlodq) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fb3b2fc9ab4ef868f74a0d0070cd8e5d5d3b247/pkg/sql/sem/tree/datum.go#L1020)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

This one will require some tests to ensure the escaping for various characters (like non-printing and valid/invalid unicode characters) matches Postgres' behaviour.
</blockquote></details>

Do I need to rewrite the format method of DByte to achieve it? I could not find a way to make use of the original serialization method to match the Postgres's format. Also, I found that for the sql query SELECT E'\\010'::bytea, they return different results.

---

*[pkg/sql/sem/tree/datum.go, line 2810 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-Kzb292Sc_Bw_bUuZ4UU-r1-2810:-L-5KJOBkiSLMbJX1fTk:b-hy5pm9) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fb3b2fc9ab4ef868f74a0d0070cd8e5d5d3b247/pkg/sql/sem/tree/datum.go#L2810)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

This doesn't seem to be tested, and I can't seem to find the corresponding docs on how this works in PG? Can you point me to it (or show me an example?)
</blockquote></details>

Sorry about it, I made a mistake. to_json method should not take DTable into consideration.

---

*[pkg/util/json/json.go, line 456 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-Kzb292Sc_Bw_bUuZ4UW-r1-456:-L-5MHDSEcs-QbEZfmiU:bwpnpa2) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fb3b2fc9ab4ef868f74a0d0070cd8e5d5d3b247/pkg/util/json/json.go#L456)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I think you might have mentioned this before but I didn't understand at the time - I think you can get away without making this function if you add a check at the end of `MakeJSON` that just returns `JSON` implementors unchanged (#20217 also adds this functionality).
</blockquote></details>

I felt that I could not catch up the comment. Could I know how to do that? Compared with MakeJSON, the pros for MakeJSONFromArrayOfJSON is that it doesn't return error and it could type check the argument, though it is somewhat verbose.

---

*[pkg/util/json/json.go, line 484 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-Kzb292Sc_Bw_bUuZ4UW-r1-484:-L-5KnaG7n3U2FLqYq-w:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fb3b2fc9ab4ef868f74a0d0070cd8e5d5d3b247/pkg/util/json/json.go#L484)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Since this function is unexported I think we can get away with a slightly less verbose name, maybe just `fromArray`?
</blockquote></details>

Done.

---

*[pkg/util/json/json.go, line 497 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-Kzb292Sc_Bw_bUuZ4UW-r1-497:-L-5KoNf_qqSSgRJPzcw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fb3b2fc9ab4ef868f74a0d0070cd8e5d5d3b247/pkg/util/json/json.go#L497)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

same here
</blockquote></details>

Done.

---

*[pkg/util/json/json.go, line 551 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-Kzb292Sc_Bw_bUuZ4UW-r1-551:-L-5Kq6nNCUXSuo6oNSY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fb3b2fc9ab4ef868f74a0d0070cd8e5d5d3b247/pkg/util/json/json.go#L551)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Nice cleanup to this function!!
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-:-L-5NLaI805s2vLdhb5S:b-240d99)*
<!-- Sent from Reviewable.io -->

justinj commented:
Review status: 0 of 5 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sem/tree/datum.go, line 277 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-Kzb292Sc_Bw_bUuZ4UU-r1-277:-L-bo_CiSM7M4-Y0kYo9:bg32e6n) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fb3b2fc9ab4ef868f74a0d0070cd8e5d5d3b247/pkg/sql/sem/tree/datum.go#L277)):*
<details><summary><i>Previously, heyihong (Yihong He) wrote…</i></summary><blockquote>

Adding more switch cases will be more convenient to use. But I am concerning that using MakeJSON skips type check which makes type errors a little difficult to discover after some changes and because there is no standard way to serialize time and timeofday, adding switch cases seems to give more restrictions about JSON serialization of time and timeofday.
</blockquote></details>

I could be convinced either way, I'm into keeping the interface here somewhat small but more specific methods are also fine.

---

*[pkg/sql/sem/tree/datum.go, line 1020 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-Kzb292Sc_Bw_bUuZ4UU-r1-1020:-L-bVJ-oGmCMdRcnOPtF:bjmbdbf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fb3b2fc9ab4ef868f74a0d0070cd8e5d5d3b247/pkg/sql/sem/tree/datum.go#L1020)):*
<details><summary><i>Previously, heyihong (Yihong He) wrote…</i></summary><blockquote>

Do I need to rewrite the format method of DByte to achieve it? I could not find a way to make use of the original serialization method to match the Postgres's format. Also, I found that for the sql query SELECT E'\\010'::bytea, they return different results.
</blockquote></details>

Hopefully you shouldn't have to, I think that the only difference there is that the clients (`cockroach --sql` vs. `psql`) print the values differently, if you connect to a Cockroach instance with `psql` I think the values should match up.

---

*[pkg/sql/sem/tree/datum.go, line 1044 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-L-btCxI3VaA0CstFv1T:-L-btCxK44VI33z52FHN:bb49zqo) ([raw file](https://github.com/cockroachdb/cockroach/blob/633b7095babe7b977c6f4421f3c0c6f3e6a80ea8/pkg/sql/sem/tree/datum.go#L1044)):*
> ```Go
> // Format implements the NodeFormatter interface.
> func (d *DBytes) Format(buf *bytes.Buffer, f FmtFlags) {
> 	withQuotes := f.withinArray || !f.encodeFlags.BareStrings
> ```

Can you add a comment in the commit message why this changed? I believe I understand but I'd like to verify.

---

*[pkg/util/json/json.go, line 456 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-Kzb292Sc_Bw_bUuZ4UW-r1-456:-L-bsh8f2VFzpXWO9Amy:bbw4ljg) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fb3b2fc9ab4ef868f74a0d0070cd8e5d5d3b247/pkg/util/json/json.go#L456)):*
<details><summary><i>Previously, heyihong (Yihong He) wrote…</i></summary><blockquote>

I felt that I could not catch up the comment. Could I know how to do that? Compared with MakeJSON, the pros for MakeJSONFromArrayOfJSON is that it doesn't return error and it could type check the argument, though it is somewhat verbose.
</blockquote></details>

Ok, I see your argument. I don't have strong feelings either way.

---

*[pkg/util/json/json.go, line 484 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-Kzb292Sc_Bw_bUuZ4UW-r1-484:-L-bqsYU53L0AlJOm9UT:bcvhpzf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fb3b2fc9ab4ef868f74a0d0070cd8e5d5d3b247/pkg/util/json/json.go#L484)):*
<details><summary><i>Previously, heyihong (Yihong He) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

`fromArray` ?

---

*[pkg/util/json/json.go, line 497 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-Kzb292Sc_Bw_bUuZ4UW-r1-497:-L-brvOJjrDEPw48bBgS:b8lxvft) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fb3b2fc9ab4ef868f74a0d0070cd8e5d5d3b247/pkg/util/json/json.go#L497)):*
<details><summary><i>Previously, heyihong (Yihong He) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

`FromMap`

---

*[pkg/util/json/json.go, line 456 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-L-bsems2oEvJsb6svx-:-L-bsems2oEvJsb6svx0:bg84boy) ([raw file](https://github.com/cockroachdb/cockroach/blob/633b7095babe7b977c6f4421f3c0c6f3e6a80ea8/pkg/util/json/json.go#L456)):*
> ```Go
> 
> // MakeJSONFromNumber returns a JSON value given a json.Number.
> func MakeJSONFromNumber(v json.Number) (JSON, error) {
> ```

`FromNumber`

---

*[pkg/util/json/json.go, line 466 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-L-bsdMVN-Kp8lbC9jy0:-L-bsdMVN-Kp8lbC9jy1:bi6toq2) ([raw file](https://github.com/cockroachdb/cockroach/blob/633b7095babe7b977c6f4421f3c0c6f3e6a80ea8/pkg/util/json/json.go#L466)):*
> ```Go
> 
> // MakeJSONFromString returns a JSON value given a string.
> func MakeJSONFromString(v string) JSON {
> ```

`FromString`

---

*[pkg/util/json/json.go, line 471 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-L-bscF7Se2PeO9KsYLI:-L-bscF7Se2PeO9KsYLJ:b-he2ydr) ([raw file](https://github.com/cockroachdb/cockroach/blob/633b7095babe7b977c6f4421f3c0c6f3e6a80ea8/pkg/util/json/json.go#L471)):*
> ```Go
> 
> // MakeJSONFromBool returns a JSON value given a bool.
> func MakeJSONFromBool(v bool) JSON {
> ```

`FromBool`

---

*[pkg/util/json/json.go, line 513 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-L-bsGOxq4E_2yu32IfD:-L-bsGOyFFOS8J1vCBLO:b8lvl86) ([raw file](https://github.com/cockroachdb/cockroach/blob/633b7095babe7b977c6f4421f3c0c6f3e6a80ea8/pkg/util/json/json.go#L513)):*
> ```Go
> 
> // MakeJSONFromInt returns a JSON value given a int.
> func MakeJSONFromInt(v int) JSON {
> ```

`FromInt`

---

*[pkg/util/json/json.go, line 520 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-L-bsHP19AI1HXAPhxm-:-L-bsHP19AI1HXAPhxm0:bwf6g1k) ([raw file](https://github.com/cockroachdb/cockroach/blob/633b7095babe7b977c6f4421f3c0c6f3e6a80ea8/pkg/util/json/json.go#L520)):*
> ```Go
> 
> // MakeJSONFromInt64 returns a JSON value given a int64.
> func MakeJSONFromInt64(v int64) JSON {
> ```

`FromInt64`

---

*[pkg/util/json/json.go, line 527 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-L-bsISNIyPQktO061xZ:-L-bsISNIyPQktO061x_:b-tamfp) ([raw file](https://github.com/cockroachdb/cockroach/blob/633b7095babe7b977c6f4421f3c0c6f3e6a80ea8/pkg/util/json/json.go#L527)):*
> ```Go
> 
> // MakeJSONFromFloat64 returns a JSON value given a float64.
> func MakeJSONFromFloat64(v float64) (JSON, error) {
> ```

`FromFloat64`

---

*[pkg/sql/sem/tree/as_json.go, line 57 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-L-5KyhV1rnJ6J1uLeQ3:-L-bsUj_gD4Pu-M_d_ZC:b4zq4wk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4aa2f81c45f95baf15fb784c3639b8cfc86fd80/pkg/sql/sem/tree/as_json.go#L57)):*
<details><summary><i>Previously, heyihong (Yihong He) wrote…</i></summary><blockquote>

I should return an error at the end of the code. Could I return like this pgerror.NewErrorf(pgerror.CodeInternalError, "unexpected datum type %T", d.(type))?
</blockquote></details>

I think you're commenting on a file that no longer exists, but that error looks fine to me.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234)*
<!-- Sent from Reviewable.io -->

heyihong commented:
Review status: 0 of 5 files reviewed at latest revision, 13 unresolved discussions.

---

*[pkg/sql/sem/tree/datum.go, line 1044 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-L-btCxI3VaA0CstFv1T:-L-cutFLzgFm_kdy9wc8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/633b7095babe7b977c6f4421f3c0c6f3e6a80ea8/pkg/sql/sem/tree/datum.go#L1044)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Can you add a comment in the commit message why this changed? I believe I understand but I'd like to verify.
</blockquote></details>

Done.

---

*[pkg/util/json/json.go, line 456 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-Kzb292Sc_Bw_bUuZ4UW-r1-456:-L-csKVvkRedk7ke5Tnf:b-m1h1s2) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fb3b2fc9ab4ef868f74a0d0070cd8e5d5d3b247/pkg/util/json/json.go#L456)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Ok, I see your argument. I don't have strong feelings either way.
</blockquote></details>

I changed its name to FromArrayOfJSON

---

*[pkg/util/json/json.go, line 484 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-Kzb292Sc_Bw_bUuZ4UW-r1-484:-L-csGY4SklGd-GBB3f8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fb3b2fc9ab4ef868f74a0d0070cd8e5d5d3b247/pkg/util/json/json.go#L484)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

`fromArray` ?
</blockquote></details>

Done.

---

*[pkg/util/json/json.go, line 497 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-Kzb292Sc_Bw_bUuZ4UW-r1-497:-L-csG0WzknO2huRjS5V:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fb3b2fc9ab4ef868f74a0d0070cd8e5d5d3b247/pkg/util/json/json.go#L497)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

`FromMap`
</blockquote></details>

Done.

---

*[pkg/util/json/json.go, line 456 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-L-bsems2oEvJsb6svx-:-L-csILpVf2vAVSE8Xma:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/633b7095babe7b977c6f4421f3c0c6f3e6a80ea8/pkg/util/json/json.go#L456)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

`FromNumber`
</blockquote></details>

Done.

---

*[pkg/util/json/json.go, line 466 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-L-bsdMVN-Kp8lbC9jy0:-L-csHX-EKyzc692QsFV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/633b7095babe7b977c6f4421f3c0c6f3e6a80ea8/pkg/util/json/json.go#L466)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

`FromString`
</blockquote></details>

Done.

---

*[pkg/util/json/json.go, line 471 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-L-bscF7Se2PeO9KsYLI:-L-csH3loZCq-cuJ8TGJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/633b7095babe7b977c6f4421f3c0c6f3e6a80ea8/pkg/util/json/json.go#L471)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

`FromBool`
</blockquote></details>

Done.

---

*[pkg/util/json/json.go, line 513 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-L-bsGOxq4E_2yu32IfD:-L-csEIEJMlbix9xI7Vq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/633b7095babe7b977c6f4421f3c0c6f3e6a80ea8/pkg/util/json/json.go#L513)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

`FromInt`
</blockquote></details>

Done.

---

*[pkg/util/json/json.go, line 520 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-L-bsHP19AI1HXAPhxm-:-L-csEpjeHPix80i3uBB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/633b7095babe7b977c6f4421f3c0c6f3e6a80ea8/pkg/util/json/json.go#L520)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

`FromInt64`
</blockquote></details>

Done.

---

*[pkg/util/json/json.go, line 527 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-L-bsISNIyPQktO061xZ:-L-csDhIq07Xzuy2LASd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/633b7095babe7b977c6f4421f3c0c6f3e6a80ea8/pkg/util/json/json.go#L527)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

`FromFloat64`
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/as_json.go, line 57 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-L-5KyhV1rnJ6J1uLeQ3:-L-csDN1nlYwZPhzuSND:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4aa2f81c45f95baf15fb784c3639b8cfc86fd80/pkg/sql/sem/tree/as_json.go#L57)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I think you're commenting on a file that no longer exists, but that error looks fine to me.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234)*
<!-- Sent from Reviewable.io -->

justinj commented:
This looks good to me now! I'd like @knz to take a look since he expressed interest before.

heyihong commented:
Review status: 0 of 7 files reviewed at latest revision, 15 unresolved discussions, some commit checks pending.

---

*[pkg/sql/sem/tree/datum.go, line 3058 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-L-gUuC9WWEAEX6WC96Z:-L-iAXSx6yODI-n_6PK8:b68pufq) ([raw file](https://github.com/cockroachdb/cockroach/blob/97a8f032118fdcb8e5a45e959cdf9dcdc123c63b/pkg/sql/sem/tree/datum.go#L3058)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I don't understand why this method needs to be implemented in the `tree` package. It is only used by `builtins.go`, why can't it be implemented in the `builtins` package?
</blockquote></details>

Is it ok to expose dNull? If it is, I will move this function to builtins package.

---

*[pkg/sql/sem/tree/datum.go, line 3089 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-L-gUjZz7GNLBZOqeHyv:-L-iAJbZKp-4kDwfXHpY:brkiuov) ([raw file](https://github.com/cockroachdb/cockroach/blob/97a8f032118fdcb8e5a45e959cdf9dcdc123c63b/pkg/sql/sem/tree/datum.go#L3089)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

What about `DTuple`? 
</blockquote></details>

Sorry for my negligence. I have added its implementation

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234)*
<!-- Sent from Reviewable.io -->

heyihong commented:
Review status: 0 of 8 files reviewed at latest revision, 15 unresolved discussions, some commit checks pending.

---

*[pkg/sql/sem/tree/datum.go, line 3058 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-L-gUuC9WWEAEX6WC96Z:-L-jAyl5vJU0jXPO4j_d:b-sy6fpj) ([raw file](https://github.com/cockroachdb/cockroach/blob/97a8f032118fdcb8e5a45e959cdf9dcdc123c63b/pkg/sql/sem/tree/datum.go#L3058)):*
<details><summary><i>Previously, heyihong (Yihong He) wrote…</i></summary><blockquote>

Is it ok to expose dNull? If it is, I will move this function to builtins package.
</blockquote></details>

Done. Instead of checking type, I found that it is able to check value to figure out whether the type of datum is dNull.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234)*
<!-- Sent from Reviewable.io -->

knz commented:
Looks good, but Justin please have a last look to check the various additions around KV pairs.

---

Reviewed 8 of 8 files at r5.
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-:-L-lYD7sZBmVpTlvep_t:bjzq6hn)*
<!-- Sent from Reviewable.io -->

justinj commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -3188,3 +3205,49 @@ func truncateTimestamp(
+			if err != nil {
+				return nil, err
+			}
+			kvPairs[i] = json.MakeKeyValuePair(fmt.Sprintf("f%d", i+1), j)
```
I think rather than exposing this extra stuff, I'd prefer to just construct a `map` and then use `MakeJSON`, especially since this is a pretty niche use-case, and I think having fewer exposed methods of creating JSON values is preferable.

heyihong committed:
sql: support to_json of JSON builtins

Release note: added asJSON function for converting Datum to JSON. Changed Format method of DBytes to fit the format requirement in sql logictest, since the string format of DBytes should not have quotes when it is not in array and its encode flags contain BareStrings.

heyihong commented:
Review status: 3 of 5 files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---

*[pkg/sql/sem/builtins/builtins.go, line 3242 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234#-Kzb292Sc_Bw_bUuZ4UT-r5-3242:-L05kO25fosl8DfgkaTx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c998ccba8e1b596700c1c70895de31d6ffd1bd4/pkg/sql/sem/builtins/builtins.go#L3242)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I think rather than exposing this extra stuff, I'd prefer to just construct a `map` and then use `MakeJSON`, especially since this is a pretty niche use-case, and I think having fewer exposed methods of creating JSON values is preferable.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20234)*
<!-- Sent from Reviewable.io -->

justinj commented:
Looks great, thanks for doing this work!
